### PR TITLE
add conformer

### PR DESCRIPTION
**Pull request by ParkJinHyeock**
conformer를 add하였습니다. https://github.com/TensorSpeech/TensorFlowASR/blob/main/tensorflow_asr/models/conformer.py 
git을 참고하였는데,  relative position encoding을 구현하려면 multi head attention 내부를 상속받아서 수정해야할 것 같아서 일단 놔뒀습니다. (저 구현에서도 relative를 사용할 지 말지 정하는 옵션이 있어 일단 사용하지 않아도 문제 없다고 생각했습니다)

https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention

그리고 layer를 encoder로 바꿨습니다.

### CODE REVIEW

File: modules.py [hunk 1]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+            kernel_regularizer=kernel_regularizer)(conv)
+
+        conv = tf.reshape(conv, [-1, time, emb])
+        conv = Dropout(dropout_rate)(conv)
+        conv = conv + x
+
+        # FFN
+        ffn = LayerNormalization()(x)
+        ffn = Dense(4*emb, activation=activation)(x)
+        ffn = Dropout(dropout_rate)(ffn)
```
Comment (lsg1213) [inline modules.py:476]: 제가 보기에 여기 입력이 뭔가 잘못되어 있는 것 같은데 확인 부탁드립니다. 위쪽의 conv가 계산되었는데 아래로 내려오지 않는 것 같고 475번째줄에도 ffn이 입력이 들어가야 할 것 같은데 제가 혹시 해석을 잘못 한 게 아닌지 궁금합니다.

Comment (lsg1213) [inline modules.py]: 제가 보기에 여기 입력이 뭔가 잘못되어 있는 것 같은데 확인 부탁드립니다. 위쪽의 conv가 계산되었는데 아래로 내려오지 않는 것 같고 475번째줄에도 ffn이 입력이 들어가야 할 것 같은데 제가 혹시 해석을 잘못 한 게 아닌지 궁금합니다.

File: modules.py [hunk 2]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+            inputs =  force_1d_inputs()(inputs)
+        x = inputs
+        batch, time, emb = x.shape
+
+        # FFN Modules
+        ffn = LayerNormalization()(x)
+        ffn = Dense(4*emb, activation=activation)(ffn)
+        ffn = Dropout(dropout_rate)(ffn)
+        ffn = Dense(emb)(ffn)
+        ffn = Dropout(dropout_rate)(ffn)
```
Comment (lsg1213) [inline modules.py]: ffn module이 반복되는 것 같은데 이게 혹시 이 모듈 외에 다른 모듈에서도 사용이 될 경우가 있을까요?

File: modules.py (hunk 1 above)
Comment (ParkJinHyeock) [reply modules.py]: 그러네요 ㅠ 코드를 복사하다가 놓친 것 같네요

File: modules.py (hunk 2 above)
Comment (ParkJinHyeock) [reply modules.py]: 음 한 번 본 것 같기는한데 여러개 중에서 세번 봤습니다. 만약 몇 개 더 있다면 이걸 일반화해도 좋을 것 같습니다.

File: modules.py [hunk 3]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
         return inputs
     return identity
 
+
+def conformer_encoder_block(model_config: dict):
+    # mandatory parameters
+    key_dim = model_config.get('key_dim', 36)
+    n_head = model_config.get('n_head', 4)
+    kernel_size = model_config.get('kernel_size', 32) # 32 
+    activation = model_config.get('activation', 'swish')
```
Comment (daniel03c1) [inline modules.py:400]: tf에서 swish를 이렇게 지원하는지 덕분에 알게됐습니다.
혹시 activation 쪽도 search space에 넣는게 좋을까요?

File: modules.py [hunk 4]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+        pos_encoding = basic_pos_encoding
+    elif pos_encoding == 'rff': # random fourier feature
+        pos_encoding = rff_pos_encoding
+    else:
+        pos_encoding = None
+    
+    def conformer_block(inputs):
+        
+        if len(inputs.shape) == 4:
+            inputs =  force_1d_inputs()(inputs)
```
Comment (daniel03c1) [inline modules.py]: force_1d_inputs()는 설계상 어떤 입력이 들어와도 자동으로 1D로 바꿔주기 때문에 
if문 없이 그냥 사용하셔도 좋을것 같습니다.

File: modules.py [hunk 5]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+        x = inputs
+        batch, time, emb = x.shape
+
+        # FFN Modules
+        ffn = LayerNormalization()(x)
+        ffn = Dense(4*emb, activation=activation)(ffn)
+        ffn = Dropout(dropout_rate)(ffn)
+        ffn = Dense(emb)(ffn)
+        ffn = Dropout(dropout_rate)(ffn)
+        x = x + 0.5*ffn
```
Comment (daniel03c1) [inline modules.py]: 원본 코드에는 fc_factor로 되어 있는것을 0.5로 적으신 것 같은데, 마찬가지로 fc_factor로 해서
추상화시키는건 어떤가요?

File: modules.py (hunk 2 above)
Comment (daniel03c1) [reply modules.py]: 첫번째 LayerNormalization을 제외하고, Dense+activation, dropout, dense, dropout은 자주쓰이는 패턴인데,
transformer_encoder_block에도 그 패턴이 들어가거든요. 거기랑 같이 묶어서 저 4개를 일반화하는거는 어떨까요?
그리고 Dense를 Conv1D로 바꿔서 조금 더 일반화하면 더 좋지 않을까 생각이 됩니다.

File: modules.py [hunk 6]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+
+
```
Comment (daniel03c1) [inline modules.py]: 함수 안에서는 한줄만 띄는게 좋을 것 같습니다!

File: modules.py [hunk 7]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+        x = attn + x
+
+
+        # Conv Block
+        conv = tf.reshape(x, [-1, time, 1, emb])
+        
+        conv = LayerNormalization()(conv)
+        conv = Conv2D(filters=2 * emb, 
+                      kernel_size=1, strides=1, padding='valid',
+                      kernel_regularizer=kernel_regularizer,)(conv)
```
Comment (daniel03c1) [inline modules.py]: 구현된 코드에서는 2D를 사용했는데, 사실상 kernel size를 1로 해서, 1D랑 동일한 기능인데
2D로 차원 추가해서, 작업하는게 1D에서 작업하는거랑 동일한 작업을 더 복잡하게 하는것 같은데
Conv1D로 바꾸는건 어떤가요? 논문에서도 어짜피 1D로 작업하는것 같아서요.

Comment (daniel03c1) [reply modules.py]: 여기서 2*emb라고 되어 있는데 이런 부분도 추상화해서 밖으로 빼면 더 좋을 것 같습니다.
그리고 padding='valid'로 되어 있는데,  지금 same으로 바꿔도 차이가 없는데, 
혹시나 kernel_size 조절하게 되면, 'same'으로 바꾸는게 더 좋지 않을까 생각이 듭니다.

Comment (daniel03c1) [reply modules.py]: 아 2가 나중에 gate때문이었군요 그러면 괜찮을 것 같습니다.
그리고 지금은 conv_1, conv_2 둘다 emb만큼의 채널을 갖게 되는데, 일반 convolution처럼
여기서 multiplier를 넣는건 어떤가요? 즉 bottleneck 구조처럼 더 커질수도 작아질수도 있게 일반화하는거죠

File: modules.py [hunk 8]
```diff
@@ -395,3 +395,90 @@ def identity(inputs):
...
+        
+        # GLU Part
+        conv_1, conv_2 = tf.split(conv, 2, axis=-1)
+        conv_2 = tf.keras.activations.sigmoid(conv_2)
+        conv = tf.multiply(conv_1, conv_2)        
+        conv = DepthwiseConv2D(kernel_size=(kernel_size, 1), 
+                               strides=1, padding="same", 
+                               kernel_regularizer=kernel_regularizer)(conv)
```
Comment (daniel03c1) [inline modules.py]: Conv1D에서 groups = channel가 되면 depthwise가 되는데, Conv1D로 하고 groups로 조절하는건 어떤가요?

**Comment by daniel03c1** [comment]
제가 transformer encoder의 변형 모듈들 예시중 하나로 conformer를 들었는데요, 혹시 다른 모듈들은 아직 보시지는 않으신거죠?

File: modules.py (hunk 3 above)
Comment (ParkJinHyeock) [reply modules.py:400]: 음 개인적인 경험에 비추어봤을 때, Swish, selu, elu, gelu, relu 를 바꿔서 써봤는데 크게 차이는 없었습니다. 물론 차이가 있긴 하겠지만 종류에 비해서 크게 효용성이 있을지는 약간 의문이긴합니다..

File: modules.py (hunk 5 above)
Comment (ParkJinHyeock) [reply modules.py]: 음 사실 논문에서 1/2로 고정해서 사용하였기 때문에 이렇게 사용하였는데요, 사실 왜 1/2을 사용하는지 몰라서 튜닝하기가 살짝 애매해서 그냥 휴리스틱하게 바꾸었습니다. 그래도 그냥 default 1/2로 두는게 더 좋을 것 같습니다

File: modules.py (hunk 2 above)
Comment (ParkJinHyeock) [reply modules.py]: Conv 1D로 하려면 kernel size를 length로 하고 filter size를 다음 linear unit으로 해서 구현하면 될까요..? 생각을 해본적이없는데 혹시 조금 더 구체적으로 말씀해주실 수 있을까요?

File: modules.py (hunk 8 above)
Comment (ParkJinHyeock) [reply modules.py]: 오우 좋은 아이디어네요

File: modules.py (hunk 7 above)
Comment (ParkJinHyeock) [reply modules.py]: 사실 이렇게 나누는 이유가 GLU라는 layer의 정의를 구현하기 위함인데, 
더 키울 이유가 있을까요? 음.. 한 채널에만 비선형을 주고 곱하는건데  sigmoid를 더 많이준다음 곱하는건 크게 의미가 없을 것 같고, 그렇다고 다른 비선형 함수를 넣은다음 곱하는게 유의미한 서치스페이스를 찾을 수 있을지는 약간 의문이기는 해요. 
GLU 를 좀 더 크게 키우는게 의미가 있다는 사례나 논리가 있으면 괜찮을텐데 아직은 약간 애매한 것 같습니다.

File: modules.py (hunk 3 above)
Comment (daniel03c1) [reply modules.py:400]: 넵 알겠습니다! activation은 고정으로 하는걸로 알고 진행하겠습니다!

File: modules.py (hunk 2 above)
Comment (daniel03c1) [reply modules.py]: Conv1D에서 kernel size가 1이면 Dense와 동일하게 작동되잖아요 그렇게 하면 되지 않을까 싶어서요

**Comment by ParkJinHyeock** [comment]
@daniel03c1 음 찾아보고 있는데  적용해볼만한게 많지는 않네요. https://arxiv.org/pdf/2103.14030.pdf 이번에 새로나온 논문이 있긴한데 아직 논문 내용이 완전히 이해가 간게 아닌데 attention을 할때 그냥 일자로 펴는게 아니라 윈도우 위치를 바꿔가면서 쓰는 것 같은데 조금 더 읽어봐야할 것 같아서요

File: modules.py (hunk 5 above)
Comment (daniel03c1) [reply modules.py]: 넵 알겠습니다!

File: modules.py (hunk 7 above)
Comment (daniel03c1) [reply modules.py]: 제가 이해한건 gate 만들어서 하나는 입력채널에 맞는 크기로 conv하고, 
또 하나도 입력채널에 맞는 크기로 conv하는데 여기에 sigmoid를 해서 미리 계산한 conv에 gate역할을 하도록 하는걸로
이해했는데 제가 이해한게 맞나요?

저는 이거를 conv1 * sigmoid(conv2) 이렇게 이해해서, 반드시 입력채널에 filters를 맞출필요가 없다고 생각해서 제안해봤습니다.
크게 중요하진 않아서 진혁씨 말대로 지금처럼 진행해도 괜찮을 것 같습니다.

File: modules.py [hunk 9]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+
+def conformer_encoder_block(model_config: dict):
+    # mandatory parameters
+    key_dim = model_config.get('key_dim', 36)
+    n_head = model_config.get('n_head', 4)
+    kernel_size = model_config.get('kernel_size', 32) # 32 
+    activation = model_config.get('activation', 'swish')
+    dropout_rate = model_config.get('dropout_rate', 0.1)
+    multiplier = model_config.get('multiplier', 4)
+    ffn_factor = model_config.get('ffn_factor', 0.2)
```
Comment (daniel03c1) [inline modules.py]: 혹시 예전에 0.5 아니었나요? 0.2를 의도하신건거죠?

File: modules.py [hunk 10]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+                                  key_dim,
+                                  dropout=dropout_rate)(attn, attn)
+        attn = Dropout(dropout_rate)(attn)
+        x = attn + x
+
+        # Conv Module
+        conv = LayerNormalization()(x)
+        conv = Conv1D(filters=2*emb, 
+                      kernel_size=1,
+                      strides=1, padding='same',
```
Comment (daniel03c1) [inline modules.py]: kernel size는 무조건 1인 케이스니까 strides, padding은 빼는건 어떤가요?

File: modules.py [hunk 11]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+                      kernel_regularizer=kernel_regularizer,)(conv)
```
Comment (daniel03c1) [inline modules.py]: 코딩 스타일 관련인데, kernel_regularizer=kernel_regularizer,) 이렇게 괄호 바로 앞에 ,가 있는데
)(conv)를 다음줄에 넘기거나, 콤마를 없애는게 더 좋지 않을까 싶습니다!

File: modules.py [hunk 12]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+        
+        # GLU Part
+        conv_1, conv_2 = tf.split(conv, 2, axis=-1)
+        conv_1 = Dense(emb)(conv_1)
+        conv_2 = Dense(emb)(conv_2)
+        conv_2 = tf.keras.activations.sigmoid(conv_2)
```
Comment (daniel03c1) [inline modules.py]: conv_2에서 바로 activation을 sigmoid로 넣는건 어떤가요?

Comment (daniel03c1) [reply modules.py]: conv_2 = Dense(emb, activation='sigmoid')(conv_2) 이런식으로요

File: modules.py [hunk 13]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+        conv = tf.multiply(conv_1, conv_2)
```
Comment (daniel03c1) [inline modules.py]: 혹시 tf.multiply(conv_1, conv_2)랑 conv_1 * conv_2랑 다른 결과가 나오나요?

File: modules.py [hunk 14]
```diff
@@ -391,3 +391,94 @@ def identity(inputs):
...
+                      groups=emb, 
+                      strides=1, padding="same", 
+                      kernel_regularizer=kernel_regularizer)(conv)
+
+        conv = BatchNormalization()(conv)
+        conv = tf.keras.activations.swish(conv) 
+        conv = Conv1D(filters=emb, 
+                      kernel_size=1, 
+                      strides=1, 
+                      padding="same",
```
Comment (daniel03c1) [inline modules.py]: 위에도 언급하긴했지만, kernel_size=1가 고정일 것 같은데, strides, padding은 따로 안하지 않아도 괜찮지 않을까 싶습니다

File: modules_test.py [hunk 15]
```diff
@@ -215,6 +215,25 @@ def test_identity_block(self):
...
+        model_config = {
+            'key_dim': 36, # mandatory
+            'n_head' : 4,
+            'kernel_size' : 32,
+            'activation': 'swish',
+            'dropout_rate': 0,
+        }
+    
+        dropout_rate = model_config.get('dropout_rate', 0.1)
+
```
Comment (daniel03c1) [inline modules_test.py]: 혹시 여기 dropout_rate의 용도가 어떻게 되는지 알 수 있을까요?
사용처가 보이지 않는데, 코드가 남아있는 것 같습니다.
